Fragments replace macros

boutell opened the pull request:
For those just tuning in: oops, Nunjucks macros don't support async. Areas in A3 require async support. 😂

We can't fix this in Nunjucks core in a reasonable amount of time, but we can provide a roughly equal parallel feature, which we're calling fragments. We also plan to actively deprecate macros in A3 code so developers don't get confused.

This is an iterative step, but it's a large subset of what we want, so I think it's time to PR it in. Things we've got:

* It works!
* Good syntax, as close to macro syntax as is possible (note the new "render" tag in the example, required because `{{ ... }}` is synchronous)
* "Import... as" works. We don't need a separate keyword for it 🎉

Things we're following up on later:

* "with context" is not implemented
* Default values for fragment parameters are not implemented
* "macro" should throw errors or deprecation warnings, if practical

Example of usage:

```
{# in areas.html #}
{% fragment main(page) %}
  {% area page, 'main' %}
{% endfragment %}

{% fragment below(page) %}
  {% area page, 'below' %}
{% endfragment %}

{# in page.html #} 

{% import "areas.html" as areas %}

{% block main %}
  <div class="home">
    <div class="content">
      {% render areas.main(data.page) %}
      {% render areas.below(data.page) %}
    </div>
  </div>
{% endblock %}
```

As with macros, you can also render a fragment defined in the same file, in which case you just call it by name (no namespace is required). Fragments with names starting with `_` are never exported from imported files, just like macros.

boutell commented:
Fragments can be passed to other fragments, in the manner of passing macros to other macros. If you've never stumbled on a use case for this it sounds strange, but people do it, so it's good that it works.

abea reviewed (commented):
This is mostly Nunjucks tag syntax that I'm not familiar with. Could we add some tests to help with confirming this?

boutell commented:
Fair

On Fri, Feb 5, 2021 at 12:05 PM Alex Bea <notifications@github.com> wrote:

> *@abea* commented on this pull request.
>
> This is mostly Nunjucks tag syntax that I'm not familiar with. Could we
> add some tests to help with confirming this?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/apostrophecms/apostrophe/pull/2711#pullrequestreview-584533154>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAH27MK2E7FJO3QOJXP2VDS5QQMRANCNFSM4XDTULWQ>
> .
>


-- 

THOMAS BOUTELL | CHIEF TECHNOLOGY OFFICER
APOSTROPHECMS | apostrophecms.com | he/him/his

myovchev commented:
Macros are the real nunjucks power, and yes I'm re-referencing macros from inside of macros sometimes :)
I'm going to follow that because it's really close related to my attempts for apos/nunjucks own design system. The end goal of it is to show how your components (mostly macros) are related (atomic design - atoms, molecules, organisms). Additionally I really want to introduce automatic documentation (e.g. from nunjuck comments) of the macro function arguments (similar to what storybook does for React components). Now that you might have control on the parser, we can collaborate on that matter :)

A quick concern - I don't think deprecating/removing macro support is a good thing. Most of the time my macros are synchronous. Is it absolutely necessary to go that route? I'll gladly mix fragments and macros in my projects.

boutell commented:
Our thinking is that if our fragments can do pretty much anything macros
can, and users are going to "stub their toe" and get confused every time
areas just mysteriously don't work in a macro, then we really should
deprecate macros so it is not confusing.

On Sun, Feb 7, 2021 at 1:06 PM Miro Yovchev <notifications@github.com>
wrote:

> Macros are the real nunjucks power, and yes I'm re-referencing macros from
> inside of macros sometimes :)
> I'm going to follow that because it's really close related to my attempts
> for apos/nunjucks own design system. The end goal of it is to show how your
> components (mostly macros) are related (atomic design - atoms, molecules,
> organisms). Additionally I really want to introduce automatic documentation
> (e.g. from nunjuck comments) of the macro function arguments (similar to
> what storybook does for React components). Now that you might have control
> on the parser, we can collaborate on that matter :)
>
> A quick concern - I don't think deprecating/removing macro support is a
> good thing. Most of the time my macros are synchronous. Is it absolutely
> necessary to go that route? I'll gladly mix fragments and macros in my
> projects.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/apostrophecms/apostrophe/pull/2711#issuecomment-774723284>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAH27LRKRIBY53OVDWZZLLS53JCPANCNFSM4XDTULWQ>
> .
>


-- 

THOMAS BOUTELL | CHIEF TECHNOLOGY OFFICER
APOSTROPHECMS | apostrophecms.com | he/him/his

boutell commented:
There is a small performance hit, yes, but that might be worth it to avoid
confusion on this point.

On Sun, Feb 7, 2021 at 4:34 PM Tom Boutell <tom@apostrophecms.com> wrote:

> Our thinking is that if our fragments can do pretty much anything macros
> can, and users are going to "stub their toe" and get confused every time
> areas just mysteriously don't work in a macro, then we really should
> deprecate macros so it is not confusing.
>
> On Sun, Feb 7, 2021 at 1:06 PM Miro Yovchev <notifications@github.com>
> wrote:
>
>> Macros are the real nunjucks power, and yes I'm re-referencing macros
>> from inside of macros sometimes :)
>> I'm going to follow that because it's really close related to my attempts
>> for apos/nunjucks own design system. The end goal of it is to show how your
>> components (mostly macros) are related (atomic design - atoms, molecules,
>> organisms). Additionally I really want to introduce automatic documentation
>> (e.g. from nunjuck comments) of the macro function arguments (similar to
>> what storybook does for React components). Now that you might have control
>> on the parser, we can collaborate on that matter :)
>>
>> A quick concern - I don't think deprecating/removing macro support is a
>> good thing. Most of the time my macros are synchronous. Is it absolutely
>> necessary to go that route? I'll gladly mix fragments and macros in my
>> projects.
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/apostrophecms/apostrophe/pull/2711#issuecomment-774723284>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AAAH27LRKRIBY53OVDWZZLLS53JCPANCNFSM4XDTULWQ>
>> .
>>
>
>
> --
>
> THOMAS BOUTELL | CHIEF TECHNOLOGY OFFICER
> APOSTROPHECMS | apostrophecms.com | he/him/his
>


-- 

THOMAS BOUTELL | CHIEF TECHNOLOGY OFFICER
APOSTROPHECMS | apostrophecms.com | he/him/his

myovchev commented:
It's really a big effort to migrate projects though. Any chance of making the macro deprecation/warning/error controlled by a system setting (explicitly set)? `allowNunjucksMacros: true`?

boutell commented:
Added a good test.

boutell commented:
Migrating a project to 3.x is certainly an effort already, but it is certainly possible to have a flag to disable the deprecation warning. We'll discuss it some more.

boutell commented:
Because we're hearing from both internal and external voices that disabling or deprecating macros would be too harsh, I added a linter that should find all attempts to use `area` or `component` inside a macro.

I tested it by putting this code in a template that does get loaded in `a3-demo`:

```
{% macro badMain(page) %}
  {% area page, 'main' %}
{% endmacro %}
```

This works in macro files, page templates, etc. It shouldn't matter because I added it to our nunjucks loader.

boutell commented:
@abea Can I get a review on the tests and linting I added? Hoping this is now a good first iteration that we can move forward with for now.  🙏
